### PR TITLE
Parse cmd line args

### LINKED ISSUE #5: Parse command line arguments
https://github.com/coders-school/testing/blob/master/module2/presentation_homework.md#wymagania-10-xp-za-ka%C5%BCde-spe%C5%82nione

> program (main.cpp) ma przyjmować 2 parametry z linii komend. Pierwszy to katalog, w którym będą pliki txt ze stanami gier na torach, a drugi opcjonalny to plik wyjściowy, w którym mają zostać zapisane przetworzone wyniki. Jeśli drugi parametr nie zostanie podany to wyniki mają zostać wypisane na ekran. Przykład użycia: ./bowling inputDirectory results.txt. Program oczywiście ma działać i realizować powierzone zadanie.

### LINKED ISSUE #6: Handling --help parameter in command line argument
https://github.com/coders-school/testing/blob/master/module2/presentation_homework.md#wymagania-10-xp-za-ka%C5%BCde-spe%C5%82nione

> program (main.cpp) po podaniu parametru -h lub --help ma wyświetlać krótką informację o tym co robi i jak go używać (czyli punkt powyżej)

### PR DESCRIPTION
**Pull request by niemarcin**
Close #5 ; Close #6

### CODE REVIEW

File: bowling/src/ArgumentParser.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        argv_.push_back(argv[i]);
+    }
+    appName_ = argv_[0];
+}
+
+std::string ArgumentParser::getInputDirectory() const {
+    if (argc_ < 2) {
+        return "";
+    }
+    return argv_[1];
```
Comment (kszytko) [inline bowling/src/ArgumentParser.cpp]:
```suggestion
    if (argc_ == 1) 
        return "";
    return argv_[1]; 
```

File: bowling/src/ArgumentParser.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        return "";
+    }
+    return argv_[1]; 
+}
+
+std::string ArgumentParser::getOutputFileName() const {
+    if (argc_ < 3) {
+        return "";
+    }
+    return argv_[2];
```
Comment (kszytko) [inline bowling/src/ArgumentParser.cpp]:
```suggestion
    if (argc_ <= 2) 
        return "";
    return argv_[2]; 
```
Duplicate code in `getInputDirectory`. You can extract method.

File: bowling/src/ArgumentParser.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        return "";
+    }
+    return argv_[2]; 
+}
+
+bool ArgumentParser::isHelpNeeded() const {
+     return argc_ == 1 || argc_ > 3 || cmdOptionExists("-h") || cmdOptionExists("--help");
+}
+
+bool ArgumentParser::cmdOptionExists(const std::string& option) const {
```
Comment (kszytko) [inline bowling/src/ArgumentParser.cpp]: Better name `argumentExists`

File: bowling/inc/ArgumentParser.hpp [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+class ArgumentParser {
+private:
```
Comment (kszytko) [inline bowling/inc/ArgumentParser.hpp]: private variables should be on the bottom of the class, based on C++ code style

**Review by kszytko — CHANGES_REQUESTED** [review]
Some small changes to make

File: bowling/src/ArgumentParser.cpp (hunk 2 above)
Comment (niemarcin) [reply bowling/src/ArgumentParser.cpp]: I don't know if it's worth to extract a method here. Especially with your proposed changes and different comparison operators: `==` in `getInputDirectory()` and `<=` in `getOutputFileName()`. I'd rather stick with simple `if - return`  statement as it is.

File: bowling/inc/ArgumentParser.hpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+public:
+    ArgumentParser(int argc, char** argv);
+
+    std::string getAppName() const { return appName_; }
+    std::string getInputDirectory() const;
+    std::string getOutputFileName() const;
+    bool isHelpNeeded() const;
+    
+private:
+    int argc_;
```
Comment (kszytko) [inline bowling/inc/ArgumentParser.hpp]: If you want to parse argc into your class, make better name, like size_.

File: bowling/inc/ArgumentParser.hpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    std::vector<std::string> argv_{};
```
Comment (kszytko) [inline bowling/inc/ArgumentParser.hpp]: same here,  arguments_

File: bowling/inc/ArgumentParser.hpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    std::string appName_{};
```
Comment (kszytko) [inline bowling/inc/ArgumentParser.hpp]: Do we really need app name?

File: bowling/test/ArgumentParser-ut.cpp [hunk 8]
```diff
@@ -0,0 +1,199 @@
+#include <gtest/gtest.h>
+
+#include "ArgumentParser.hpp"
+
+TEST(ArgumentParserTests, shouldBeInitializedWithGivenAppName) {
+    // GIVEN
+    char appName[] = "./bowling";
+    char* argumentsArray[] = {appName};
+    int argumentsCounter = sizeof(argumentsArray) / sizeof(char*);
+    ArgumentParser ap{argumentsCounter, argumentsArray};
```
Comment (kszytko) [inline bowling/test/ArgumentParser-ut.cpp:10]: In all tests you have a lot of duplicated code, so you can consider to use TEST_F with predefined parameters, and some functions like `calculateArgumentsSize`

File: bowling/src/ArgumentParser.cpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        argv_.push_back(argv[i]);
+    }
+    appName_ = argv_[0];
+}
+
+std::string ArgumentParser::getInputDirectory() const {
+    if (argc_ == 1) {
+        return "";
+    }
+    return argv_[1];
```
Comment (kszytko) [inline bowling/src/ArgumentParser.cpp]: Practicly you return first argument if exists,  in `getOutputFileName` you are returning secound argument if exists.  
It's good to make it into small function like this:

```cpp
std::string getExistingArgument(size_t number) {
    if (size >= number)
        return arguments_[number - 1];
    return ""; 
}

what do you think?
```

**Review by kszytko — CHANGES_REQUESTED** [review]
Some small changes i think to make,  consider updates in tests but not so important.

File: bowling/src/ArgumentParser.cpp [hunk 10]
```diff
@@ -0,0 +1,34 @@
+#include "ArgumentParser.hpp"
+
+#include <algorithm>
+
+ArgumentParser::ArgumentParser(int argc, char** argv) {
+    argc_ = argc;
+    argv_.reserve(argc_);
+    for (int i = 0; i < argc_ ; ++i) {
+        argv_.push_back(argv[i]);
+    }
```
Comment (kszytko) [inline bowling/src/ArgumentParser.cpp]: Move it into private method `void makeVector(array)`, now it is not readable

File: bowling/src/ArgumentParser.cpp [hunk 11]
```diff
@@ -0,0 +1,34 @@
+#include "ArgumentParser.hpp"
+
+#include <algorithm>
+
+ArgumentParser::ArgumentParser(int argc, char** argv) {
```
Comment (kszytko) [inline bowling/src/ArgumentParser.cpp]:
```suggestion
ArgumentParser::ArgumentParser(int size, char** arguments)  : size_(size){
makeVector(arguments);
```

File: bowling/src/ArgumentParser.cpp [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    argc_ = argc;
+    argv_.reserve(argc_);
+    for (int i = 0; i < argc_ ; ++i) {
+        argv_.push_back(argv[i]);
+    }
+    appName_ = argv_[0];
```
Comment (kszytko) [inline bowling/src/ArgumentParser.cpp]: If you really want to save name of file,  make private method `readAppName();`

**Review by kszytko — CHANGES_REQUESTED** [review]
Some more changes to discuss

File: bowling/test/ArgumentParser-ut.cpp (hunk 8 above)
Comment (niemarcin) [reply bowling/test/ArgumentParser-ut.cpp:10]: Yes, I know. 
I'd like to refactor it later, if I had enough time for it. Until then the code is tested with current tests, so I'd like to provide the parser to other members of our group.

But you're absolutely right that TEST_F should happen here. ;)

**Review by kszytko — APPROVED** [review]
Very good code!  Only tests to future update

File: bowling/src/ArgumentParser.cpp [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+#include <algorithm>
+
+ArgumentParser::ArgumentParser(int size, char** arguments) {
+    size_ = size;
+    fillArgumentsVector(arguments);
+}
+
+void ArgumentParser::fillArgumentsVector(char** arguments) {
+    arguments_.reserve(size_);
+    for (int i = 0; i < size_; ++i) {
```
Comment (mariuszlisowski) [inline bowling/src/ArgumentParser.cpp:12]: size_t instead of int because 'size' variable is of size_t type

File: bowling/src/ArgumentParser.cpp [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+std::string ArgumentParser::getInputDirectory() const {
+    return getExistingArgument(1); 
+}
+
+std::string ArgumentParser::getOutputFileName() const {
+    return getExistingArgument(2); 
+}
+
+bool ArgumentParser::isHelpNeeded() const {
+     return size_ == 1 || size_ > 3 || argumentExists("-h") || argumentExists("--help");
```
Comment (mariuszlisowski) [inline bowling/src/ArgumentParser.cpp:30]: There's no difference if program is executed with -h option or without any arguments so the option -h seems useless (help is displayed anyway). Unless when executing without any arguments it shows only short description of its purpose and offers -h option for more details. If so the 'size_ == 1' condition is  redundant. But it's really nothing and can be left as it is.

**Review by mariuszlisowski — APPROVED** [review]

**Pull request merged**